File: change_root.py

import subprocess
import sys

def execute(string):
	proc = subprocess.Popen(string.split())
	proc.wait()
	print(string)

NUMBER = input('input integer for /dev/sda: ')
try:
	int(NUMBER)
except:
	print('You must enter an integer for sda')
	sys.exit()

one = 'sudo mount /dev/sda{} /mnt'.format(NUMBER)
two = 'sudo mount -t proc none /mnt/proc'
three = 'sudo mount -o bind /dev /mnt/dev'
four = 'sudo mount -o bind /sys /mnt/sys'
five = 'sudo chroot /mnt'

uone = 'sudo umount /mnt/sys'
utwo = 'sudo umount /mnt/dev'
uthree = 'sudo umount /mnt/proc'
ufour = 'sudo umount /mnt'

print('changing to root of /dev/sda{}'.format(NUMBER))
print('"exit" to unmount')
execute(one)
execute(two)
execute(three)
execute(four)
execute(five)

#after user enters 'exit' will return to python to unmount

print('unmounting...')
execute(uone)
execute(utwo)
execute(uthree)
execute(ufour)

File: change_root.cpp
#include <iostream>
#include <vector>

std::string MOUNTPOINT = "/mnt";

void cmd(std::string s){
    std::cout << s << std::endl;
    system(s.c_str());
    
}

void cmds(std::vector<std::string> v){
    for (unsigned int i=0; i<v.size(); i++){
        cmd(v[i]);
    }
}

int main(){
    std::string num;
    std::cout << "input integer for /dev/sda: ";
    std::cin >> num;
    
    std::cout << "***sudo grub-install /dev/sda\n***sudo update-grub" << std::endl;
    
    std::vector<std::string> v{
        "sudo mount /dev/sda" + num,
        "sudo mount -t proc none "+MOUNTPOINT+"/proc",
        "sudo mount -o bind /dev "+MOUNTPOINT+"/dev",
        "sudo mount -o bind /sys "+MOUNTPOINT+"/sys",
        "sudo chroot "+MOUNTPOINT,
        //user in mountpoint, completes after user exits
        "sudo umount "+MOUNTPOINT+"/sys",
        "sudo umount "+MOUNTPOINT+"/dev",
        "sudo umount "+MOUNTPOINT+"/proc",
        "sudo umount "+MOUNTPOINT
        
    };
    cmds(v);
}
